core/ir/riscv64/codec.py

```diff
@@ -0,0 +1,1063 @@
+        with open(template_file, 'r') as tf, open(out_file, 'w') as of:
+            for line in tf:
+                if '*/ OP_' in line:
+                    nmb = re.findall(r'\d+', line)
+                    if len(nmb) != 1:
+                        wrn(
+                            f"Template opcode line is missing an index?: '{line}'")
+                    idx = int(nmb[0])
+                elif tmpl_fld in line:
+                    found_template_fld = True
+
+                    if idx == -1:
+                        wrn("Starting index not found, using 0")
+                    idx += 1
+                    IslGenerator.OP_TBL_OFFSET = idx
+                    lines = []
+                    # Generate list of:
+                    #    OP_<opcode>,   /**< <extension> <opcode> opcode. */
+                    for i in self.instructions:
+                        lines.append(
+                            f"    /* {idx:3d} */ OP_{i.formatted_name()} = {idx},   /**< {i.ext} {i.name} opcode. */")
+                        idx += 1
+
+                    line = line.replace(tmpl_fld, '\n'.join(lines))
+                of.write(line)
+        if not found_template_fld:
+            err(f"{tmpl_fld} not found in {template_file}")
+        return found_template_fld
+
+    def generate_instr_macros(self, template_file, out_file) -> bool:
+        '''
+        Generate the instr_create_api.h header file.
+
+        Returns true if the output file was written successfully.'''
+        tmpl_fld = '@INSTR_MACROS@'
+        if len(self.instructions) == 0:
+            return False
+        found_template_fld = False
+        # Replace tmpl_fld in the template_file and save as out_file
+        with open(template_file, 'r') as tf, open(out_file, 'w') as of:
+            for line in tf:
+                if tmpl_fld in line:
+                    found_template_fld = True
+
+                    lines = []
+                    # Generate list of:
+                    #    #define INSTR_CREATE_<opcode>(dc, <arguments>) \
+                    #      instr_create_<n_dst>dst_<n_src>src(dc, OP_<opcode>, <arguments>)
+                    for i in self.instructions:
+                        flds = [f for f in i.flds]
+                        flds.reverse()
+                        args = ''
+                        arg_comments = ''
+                        if len(flds) > 0:
+                            args += ', '
+                            args += ', '.join([f.formatted_name()
+                                              for f in flds])
+                            arg_comments += '\n'
+                            arg_comments += '\n'.join(
+                                [f' * \param {f.formatted_name():6}  {f.arg_cmt}' for f in flds])
+                        nd = len([f for f in flds if f.is_dest])
+                        ns = len(flds) - nd
+                        lines.append(
+                            f'''/**
+ * Creates a {i.name} instruction.
```

Comment:
s/a/a(n)/ so nobody complains about the wrong article before a vowel?  Or I guess you could actually check whether i.name starts with a vowel.

Reply:
The problem here is that we should use "an" before a "vowel sound", not a vowel per-se (i.e. "a union") so I went with `a(n)`.